Fix `is_smooth` for projective space

paemurru opened the pull request:
Previously, taking `base_ring` from a quotient ring $S = R/I$ gave the ring $R$, but if $S$ was not a quotient ring, then it gave the base field of $S$, introducing an error in case the argument of `is_smooth` is the projective space.

paemurru committed:
Fix `is_smooth` for projective space

Previously, taking `base_ring` from a quotient ring $S = R/I$ gave the ring $R$, but if $S$ was not a quotient ring, then it gave the base field of $S$, introducing an error in case the argument of `is_smooth` is the projective space.

lgoettgens commented on src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Properties.jl:
```diff
@@ -53,7 +53,8 @@ function _jacobian_criterion(X::CoveredScheme{<:Field})
   for (V, fs) in dec_info
-    R = base_ring(OO(V))
+    R_quotient = OO(V)
+    typeof(R_quotient)<:MPolyQuoRing ? R = base_ring(R_quotient) : R = R_quotient
```
```suggestion
    R_quotient isa MPolyQuoRing ? R = base_ring(R_quotient) : R = R_quotient
```
small nitpick: this is the preferred way in julia land if I remember correctly

paemurru committed:
Update src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Properties.jl

Co-authored-by: Lars Göttgens <lars.goettgens@gmail.com>

paemurru committed:
Fix docs

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/CoveredSchemes/Objects/Properties.jl:
```diff
@@ -53,7 +53,8 @@ function _jacobian_criterion(X::CoveredScheme{<:Field})
   for (V, fs) in dec_info
-    R = base_ring(OO(V))
+    R_quotient = OO(V)
+    R_quotient isa MPolyQuoRing ? R = base_ring(R_quotient) : R = R_quotient
```
Are you 100% sure that you never see an MPolyLocalizedRing or an MPolyQuoLocalizedRing here? This is the general case of covered schemes here, not the projective scheme case. We could see for instance a complement of a hypersurface as a chart (which gives the right answer for base_ring, by the way). 

I would do this fix the other way around: 

R_quotient isa MPolyRing ? R = R_quotient : R = base_ring(R_quotient)

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -17,7 +17,7 @@ end
 @doc raw"""
-    is_smooth(P::AbsProjectiveScheme; algorithm=:default) -> Bool
+    is_smooth(P::AbsProjectiveScheme{<:Any, <:MPolyQuoRing}; algorithm=:default) -> Bool
 
```
We could rather catch the easy corner case than exclude it in the docs. 
It is better, if we have the compatibility functions to do this in one place (what you do below anyway), than several users doing it each by themselves after reading this in the docs.

simonbrandhorst replied:
A solution could be the following: Define

`is_smooth(P::AbsProjectiveScheme; algorithm=:default)`
and document just this method. Then have the following two methods
`is_smooth(P::AbsProjectiveScheme{<:any, <:MPolyQuoRing}; algorithm=:default)`
`is_smooth(P::AbsProjectiveScheme{<:any, <:MPolyRing}; algorithm=:default)`

perhaps the `any` needs to be replaced by `Field` ? ... I have little experience with algebraic geometry over general rings ... best consult a textbook to be sure.

afkafkafk13 replied:
We already have is_smooth for MPolyRing below in the new code. Your suggestion is the way to go.

Over general rings, smooth means Jacobi-criterion satisfied and regular means dim = edim. Please keep in mind that regular still allows fibre-singularities e.g. over ZZ.

afkafkafk13 reviewed (commented):
Thank you. This bug-fix is really important.

I have a few comments on details.

paemurru committed:
Allow arbitrary field for :projective_jacobian

Throw unimplemented error for :affine_cone and :covered_jacobian if base field is not a field.

paemurru committed:
Implement suggestions of Anne and Simon

paemurru committed:
Fix docstring

afkafkafk13 commented on src/AlgebraicGeometry/Schemes/ProjectiveSchemes/Objects/Properties.jl:
```diff
@@ -101,24 +122,22 @@ function is_smooth(P::AbsProjectiveScheme; algorithm=:default)
   sing_locus = ideal(R, minors(mat, codim(P)))
-  return dim(sing_locus + I) <= 0
+  irrelevant_ideal = ideal(R, gens(R))
+  return is_one(saturation(sing_locus, irrelevant_ideal))
```
I think that you want 
                       sing_locus = ideal(R, minors(mat,codim(P)) + I
Otherwise, you could also get points outside of P, because you never add generators of I in lines 139 and 140, whereas this happened in the deleted line 121.

paemurru committed:
Fix sing_locus line

Otherwise might also include points outside of P.
